File: frontend/lib/utils/shipment_delivery_note_scan.dart
List<String> deliveryNoteScanKeys(Map<String, dynamic> shipment) {
  final order = shipment['wholesale_order'] as Map<String, dynamic>?;
  final orderNumber = order?['order_number']?.toString();
  final keys = <String>[
    if (orderNumber != null && orderNumber.isNotEmpty) orderNumber,
    if (orderNumber != null && orderNumber.isNotEmpty)
      orderNumber.replaceFirst(RegExp(r'^WO-', caseSensitive: false), ''),
    if (order?['po_number'] != null) order!['po_number'].toString(),
    if (order?['ref_no'] != null) order!['ref_no'].toString(),
    shipment['id']?.toString() ?? '',
    shipment['wholesale_order_id']?.toString() ?? '',
    'shipment:${shipment['id']}',
    'shipment-${shipment['id']}',
    if (shipment['tracking_number'] != null) shipment['tracking_number'].toString(),
  ];
  final normalized = keys
      .where((k) => k.trim().isNotEmpty)
      .map((k) => k.trim().toLowerCase())
      .toSet()
      .toList();
  return normalized;
}

bool shipmentMatchesDeliveryNoteScan(Map<String, dynamic> shipment, String raw) {
  final q = raw.trim().toLowerCase();
  if (q.isEmpty) return false;
  final keys = deliveryNoteScanKeys(shipment);
  return keys.any((k) => k == q || k.contains(q) || q.contains(k));
}

bool shipmentMatchesCouriers(Map<String, dynamic> shipment, List<String> couriers) {
  if (couriers.isEmpty) return true;
  final assigned = (shipment['courier']?.toString() ?? '').trim().toLowerCase();
  if (assigned.isEmpty) return true;
  return couriers.any((c) => c.trim().toLowerCase() == assigned);
}

File: management-frontend/src/utils/shipmentDeliveryNoteScan.ts
import type { Shipment } from '../types';

function normalizeScan(raw: string): string {
  return raw.trim().toLowerCase();
}

/** Values that may appear on a delivery note barcode or label. */
export function deliveryNoteScanKeys(shipment: Shipment): string[] {
  const order = shipment.wholesale_order;
  const keys = [
    order?.order_number,
    order?.po_number,
    order?.ref_no,
    order?.order_number?.replace(/^WO-/i, ''),
    String(shipment.id),
    String(shipment.wholesale_order_id),
    `shipment:${shipment.id}`,
    `shipment-${shipment.id}`,
    shipment.tracking_number,
  ];
  return [...new Set(keys.filter((k) => k != null && String(k).trim() !== '').map((k) => normalizeScan(String(k))))];
}

export function shipmentMatchesDeliveryNoteScan(shipment: Shipment, raw: string): boolean {
  const q = normalizeScan(raw);
  if (!q) return false;
  const keys = deliveryNoteScanKeys(shipment);
  return keys.some((k) => k === q || k.includes(q) || q.includes(k));
}

export function shipmentMatchesCouriers(shipment: Shipment, couriers: string[]): boolean {
  if (couriers.length === 0) return true;
  const assigned = (shipment.courier ?? '').trim().toLowerCase();
  if (!assigned) return true;
  return couriers.some((c) => c.trim().toLowerCase() === assigned);
}
